alacritty_terminal/src/selection.rs

```diff
@@ -81,200 +82,92 @@ pub trait Dimensions {
-        match *self {
-            Selection::Simple { ref mut region } => {
-                region.end = Anchor::new(location.into(), side);
-            },
-            Selection::Semantic { ref mut region } | Selection::Lines { ref mut region, .. } => {
-                region.end = location.into();
-            },
-        }
+        // Always set side; can be ignored later if needed.
+        self.region.end = Anchor::new(location.into(), side);
     }
 
-    pub fn to_span(&self, term: &Term, alt_screen: bool) -> Option<Span> {
-        let span = match *self {
-            Selection::Simple { ref region } => Selection::span_simple(term, region, alt_screen),
-            Selection::Semantic { ref region } => {
-                Selection::span_semantic(term, region, alt_screen)
-            },
-            Selection::Lines { ref region, initial_line } => {
-                Selection::span_lines(term, region, initial_line, alt_screen)
-            },
-        };
-
-        // Expand selection across double-width cells
-        span.map(|mut span| {
-            let grid = term.grid();
-
-            if span.end.col < grid.num_cols()
-                && grid[span.end.line][span.end.col].flags.contains(Flags::WIDE_CHAR_SPACER)
-            {
-                span.end.col = Column(span.end.col.saturating_sub(1));
-            }
-
-            if span.start.col.0 < grid.num_cols().saturating_sub(1)
-                && grid[span.start.line][span.start.col].flags.contains(Flags::WIDE_CHAR)
-            {
-                span.start.col += 1;
-            }
-
-            span
-        })
-    }
-
-    pub fn is_empty(&self) -> bool {
-        match *self {
-            Selection::Simple { ref region } => {
-                region.start == region.end && region.start.side == region.end.side
-            },
-            Selection::Semantic { .. } | Selection::Lines { .. } => false,
-        }
+    pub fn update_as(&mut self, location: Point<usize>, side: Side, sel_type: SelectionType) {
+        self.update(location, side);
+        self.end_type = sel_type;
     }
 
-    fn span_semantic<G>(grid: &G, region: &Range<Point<isize>>, alt_screen: bool) -> Option<Span>
-    where
-        G: Search + Dimensions,
-    {
-        let cols = grid.dimensions().col;
-        let lines = grid.dimensions().line.0 as isize;
-
-        // Normalize ordering of selected cells
-        let (mut start, mut end) = if region.start < region.end {
-            (region.start, region.end)
-        } else {
-            (region.end, region.start)
-        };
-
-        if alt_screen {
-            Selection::alt_screen_clamp(&mut start, &mut end, lines, cols)?;
-        }
-
-        let (mut start, mut end) = if start == end {
-            if let Some(end) = grid.bracket_search(start.into()) {
-                (start.into(), end)
+    pub fn to_span(&self, term: &Term, alt_screen: bool) -> Option<Span> {
```

Comment:
This function is massive now, it should be split into smaller more compact parts.